CONTRIBUTING.md

```diff
@@ -0,0 +1,137 @@
+
+This repository uses `github.com/stretchr/testify` framework for unit tests.
+
+> TODO: to be discussed, what we mean by "significant", e.g. >70% test coverage or so.
+
+### Documentation
+
+Keeping a compelling `godoc` documentation is part of the "drop-in replacement" promise.
+
+Any newly exposed API should come with a clear, readable description of what it does.
+
+Testable example are alway welcome (although not required).
+
+Conversely, internals only need comments if a particular caveat or expectation requires wording besides just code.
+
+### Dependencies
+
+So far, `pflag` has managed to build without _any_ external dependency.
+
+We may indulge into _a few_ dependencies, especially test dependencies, or things like `golang.org/x/...`.
+
+We should not need much more than that, unless we are talking about extensions (e.g. new flag types). See below.
+
+### Commit etiquette
+
+Please keep your commit title and description concise and to the point.
+
+Feel free to `git rebase -i`, `git commit --amend` etc, so this repository keeps a nice and readable git history,
+without merge commits or fixup rescindments.
+
+### Contributor Licence Agreement
+
+Contributions are subject to a CLA DCO being signed off.
+
+> TODO: is this something that is still needed? Anyhow, check if the CLA terms are up to date.
+> The Linux Foundation has introduced [EasyCLA](https://docs.linuxfoundation.org/lfx/easycla).
+> I believe that `cla.assistant.io` is a bit old.
+
+### Signed commits
+
+Commits shall be [signed](https://docs.github.com/en/authentication/managing-commit-signature-verification/signing-commits).
+
+> TODO: this requirement is debatable.
+> On the one hand, it is certainly something many junior developers are perplex about.
+> On the other hand, supporting so many major pieces of infrastructure is a big responsibility.
+
+### Linear history
+
+Please rebase your PRs to the current master before a merging can be done. All PR's will be squashed.
+
+## Contributed extensions
+
+New types, great features built on top of `pflag` etc are welcome.
+
+Considering adding those to the already rather rich collection of supported flag types is going always to be a long process.
+
+To address that, a new repository `spf13/pflag-contrib` may host such new ideas and (possibly breaking) innovations.
+
+That repository works under basically the same set of rules, but less stringent.
+
+* backward-compatibility for a contributed feature remains desirable, but not required
+* contributors have a freer hand regarding dependencies
+* documentation & testing are needed, possibly with a lighter touch
+* introduction of new go features is possible - just guard your extension with `go: build gox.yz` 
+  to avoid breaking other contributions
+
+> I am proposing a new git repo as I found rather impractical to maintain a nested `go.mod` in the main repo.
+> Further, we most likely want to version contributed features independantly.
```

Comment:
Contrib repos are generally hard to maintain, especially as they grow over time.

I'd probably start with a list of extensions in a markdown file.

Reply:
@sagikazarmark  point taken.  Let's start with a mere document with links